internal/backend/runtime/omni/controllers/omni/cluster_workload_proxy_status.go

```diff
@@ -0,0 +1,141 @@
+				if err != nil {
+					return fmt.Errorf("failed to list cluster machine statuses: %w", err)
+				}
+
+				if !cluster.TypedSpec().Value.GetFeatures().GetEnableWorkloadProxy() {
+					if err = workloadProxyReconciler.Reconcile(cluster.Metadata().ID(), nil); err != nil {
+						return fmt.Errorf("failed to reconcile load balancers (feature disabled): %w", err)
+					}
+
+					status.TypedSpec().Value.NumExposedServices = 0
+
+					return nil
+				}
+
+				healthyTargetHosts := make([]string, 0, cmsList.Len())
+
+				for iter := cmsList.Iterator(); iter.Next(); {
+					cms := iter.Value()
+
+					if cms.Metadata().Phase() == resource.PhaseTearingDown {
+						continue
+					}
+
+					if !cms.TypedSpec().Value.GetReady() {
+						continue
+					}
+
+					if cms.TypedSpec().Value.ManagementAddress == "" {
+						continue
+					}
+
+					healthyTargetHosts = append(healthyTargetHosts, cms.TypedSpec().Value.ManagementAddress)
+				}
+
+				slices.Sort(healthyTargetHosts)
+
+				svcList, err := safe.ReaderListAll[*omni.ExposedService](ctx, r, state.WithLabelQuery(resource.LabelEqual(omni.LabelCluster, cluster.Metadata().ID())))
+				if err != nil {
+					return fmt.Errorf("failed to list exposed services: %w", err)
+				}
+
+				var errs error
+
+				aliasToUpstreamAddresses := make(map[string][]string, svcList.Len())
+
+				for iter := svcList.Iterator(); iter.Next(); {
+					svc := iter.Value()
+
+					if svc.Metadata().Phase() == resource.PhaseTearingDown {
+						continue
+					}
+
+					alias, ok := svc.Metadata().Labels().Get(omni.LabelExposedServiceAlias)
+					if !ok {
+						logger.Warn("missing label", zap.String("label", omni.LabelExposedServiceAlias), zap.String("service", svc.Metadata().ID()))
+
+						continue
+					}
+
+					aliasToUpstreamAddresses[alias] = xslices.Map(healthyTargetHosts, func(host string) string {
+						return net.JoinHostPort(host, strconv.Itoa(int(svc.TypedSpec().Value.Port)))
+					})
+				}
+
+				if err = workloadProxyReconciler.Reconcile(cluster.Metadata().ID(), aliasToUpstreamAddresses); err != nil {
+					return fmt.Errorf("failed to reconcile load balancers: %w", err)
+				}
+
+				status.TypedSpec().Value.NumExposedServices = uint32(len(aliasToUpstreamAddresses))
+
+				return errs
+			},
+			FinalizerRemovalFunc: func(_ context.Context, _ controller.Reader, logger *zap.Logger, cluster *omni.Cluster) error {
+				if err := workloadProxyReconciler.Reconcile(cluster.Metadata().ID(), nil); err != nil {
+					logger.Error("failed to reconcile load balancers (finalizer removal)", zap.Error(err))
+				}
+
+				return nil
+			},
+		},
+		qtransform.WithExtraMappedInput(
+			mappers.MapByClusterLabel[*omni.ExposedService, *omni.Cluster](),
+		),
+		qtransform.WithExtraMappedInput(
+			mappers.MapByClusterLabel[*omni.ClusterMachineStatus, *omni.Cluster](),
+		),
```

Comment:
I see you don't have tests for exposed service resource and cluster machine status deletion.
As you don't add finalizers to these two inputs the test might fail.

You can use `helpers.HandleInputs` for that, it both reads the resources and adds finalizers to them.

Reply:
Actually maybe `HandleInputs` won't fit here, but in any case there should be finalizer added to each input, which should be removed if either this additional input is tearing down or if cluster is tearing down.

Reply:
Ah, good catch, I added a test that deletes a clustermachinestatus and an exposedservice, and you were right. I modified the controller code to add and remove finalizers to those resources, and now it seems to be working fine. Please have another look.

Reply:
your fix looks good except that it doesn't remove finalizers when the cluster enters tearing down phase.
But that probably works now as cluster resource is always removed after we clean up clustermachinestatuses and exposedservice resources for it.

Reply:
Just to be on the safe side, I added the same finalizer removal logic into `FinalizerRemovalExtraOutputFunc` as well.